### PR TITLE
Some improvements to nix parsing

### PR DESCRIPTION
**Pull request by jopejoe1**
Added build logs to nix, improved the parsing of URL fields in nix, and added binary names to nix.

### CODE REVIEW

**Comment by AMDmi3** [comment]
> [add build logs to nix](https://github.com/repology/repology-updater/pull/1403/commits/f01e098ebaa6fe716660b6634791ef84ba85d044)

We need direct links to raw build log files.

> [improve url parsing in nix](https://github.com/repology/repology-updater/pull/1403/commits/06c138f14fd59a23e7e7ac528eecf16008ba3f28)

You don't need `extract_nix_url`, as `add_links` already handles argument as either str, list or None. `downloadPage` support is useful, but I'd introduce a new link type for it (repology/package.py).

> [add binary name to nix](https://github.com/repology/repology-updater/pull/1403/commits/a91285e2d9e7e8c0abe2f2d5aad74dd065d373e6)

This part is invalid, `binname` is binary package name, not executable name.

**Comment by jopejoe1** [comment]
> > [add build logs to nix](https://github.com/repology/repology-updater/pull/1403/commits/f01e098ebaa6fe716660b6634791ef84ba85d044)
> 
> We need direct links to raw build log files.



I don't think it is currently possible to get a link to the raw logs from the data that get published in the package.json as the raw log files all have hash or build ID in their URL, e.g. https://cache.nixos.org/log/s7ir5kdff5437fna8ks274sl537ks5md-firefox-devedition-unwrapped-126.0b5.drv or https://hydra.nixos.org/build/259278227/nixlog/1
 
> > [improve url parsing in nix](https://github.com/repology/repology-updater/pull/1403/commits/06c138f14fd59a23e7e7ac528eecf16008ba3f28)
> 
> You don't need `extract_nix_url`, as `add_links` already handles argument as either str, list or None. `downloadPage` support is useful, but I'd introduce a new link type for it (repology/package.py).

Would something like UPSTREAM_DOWNLOAD_PAGE be okay for that link type?
 
> > [add binary name to nix](https://github.com/repology/repology-updater/pull/1403/commits/a91285e2d9e7e8c0abe2f2d5aad74dd065d373e6)
> 
> This part is invalid, `binname` is binary package name, not executable name.

Removing it.

File: repology/parsers/parsers/nix.py [hunk 1]
```diff
@@ -182,6 +204,31 @@ def iter_parse(self, path: str, factory: PackageFactory) -> Iterable[PackageMake
                 if 'changelog' in meta:
                     pkg.add_links(LinkType.UPSTREAM_CHANGELOG, meta.get('changelog'))
 
+                for arch in ['x86_64-linux', 'aarch64-linux', 'x86_64-darwin', 'aarch64-darwin']:
```
[the remark this answers is not in the document]
Comment (AMDmi3) [reply repology/parsers/parsers/nix.py:227]: Or even better, let's do a constant map of `(arch, platform, is_trunk) -> link templates` as in

```python
_BUILD_LOGS_LINK_TEMPLATES = {
    ("x86_64", "linux", True): f'https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.{key}.x86_64-linux',
    ...
}
```

```python
for arch, platform in [('x86_64', 'linux')]:
    if nix_has_logs(meta, f'{arch}-{platform}') and (template := _BUILD_LOGS_LINK_TEMPLATES.get(arch, platform, branch is not None)):
        pkg.add_links(template.format(key = key, branch = branch))
```

which will also reduce code duplication.

File: repos.d/nixos.yaml [hunk 2]
```diff
@@ -21,6 +21,7 @@
         url: https://channels.nixos.org/nixos-{{branch}}/packages.json.br
       parser:
         class: NixJsonParser
+        branch: {{branch}}
```
Comment (AMDmi3) [inline repos.d/nixos.yaml:24]: Wrap in `{% if branch %}` and quote for proper typing.

**Review by AMDmi3 — CHANGES_REQUESTED** [review]

**Comment by AMDmi3** [comment]
> I don't think it is currently possible to get a link to the raw logs from the data that get published in the package.json as the raw log files all have hash or build ID in their URL

I've updated the the comment - let's add these links. I've requested a few changes.

> Would something like UPSTREAM_DOWNLOAD_PAGE be okay for that link type?

It would be perfect. Note that you may want to add it to a few places where UPSTREAM_HOMEPAGE or UPSTREAM_DOWNLOADS are handled, namely in `sql.d/update/update_problems.sql` of this repo and [webapp](https://github.com/repology/repology-webapp) (2 places in code and template I've just [fixed](https://github.com/repology/repology-webapp/issues/245)). If it's not apparent how these should be handled just ignore this, I'll add them myself.

File: repology/parsers/parsers/nix.py [hunk 3]
```diff
@@ -61,6 +61,23 @@ def extract_nix_licenses(whatever: Any) -> list[str]:
         return []
 
 
+def nix_has_logs(meta: dict[str, Any], arch: str) -> bool:
+    if 'platforms' not in meta or arch not in meta['platforms']:
+        return False
+    elif 'broken' in meta and meta['broken']:
```
[the remark this answers is not in the document]
Comment (jopejoe1) [reply repology/parsers/parsers/nix.py:80]: There is some documentation at https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/meta.chapter.md, but it's sadly incomplete and missing a lot of stuff.

Most of these are pretty much we know it's broken, so no need to waste resources building.

**Comment by jopejoe1** [comment]
Applied all recommend changes and add champion pr https://github.com/repology/repology-webapp/pull/246 for the webapp.

**Review by AMDmi3 — APPROVED** [review]

**Comment by AMDmi3** [comment]
Excellent, thank you!

**Pull request merged**